tarpc/src/generic_transport/mod.rs

```diff
@@ -0,0 +1,407 @@
+            }
+        }
+    }
+}
+
+impl<S, Item, ItemDecoder, SinkItem, SinkItemEncoder, SinkItemEncoderError> Sink<SinkItem>
+    for Transport<S, Item, ItemDecoder, SinkItem, SinkItemEncoder>
+where
+    S: AsyncWrite + Unpin,
+    SinkItem: Serialize,
+    SinkItemEncoder: Serializer<SinkItem>,
+    SinkItemEncoderError: Into<Box<dyn Error + Send + Sync>>,
+    FramedRead<
+        FramedWrite<Framed<S, LengthDelimitedCodec>, SinkItem, SinkItemEncoder>,
+        Item,
+        ItemDecoder,
+    >: Sink<SinkItem, Error = SinkItemEncoderError>,
+{
+    type Error = io::Error;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        convert(self.project().inner.poll_ready(cx))
+    }
+
+    fn start_send(self: Pin<&mut Self>, item: SinkItem) -> io::Result<()> {
+        self.project()
+            .inner
+            .start_send(item)
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        convert(self.project().inner.poll_flush(cx))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        convert(self.project().inner.poll_close(cx))
+    }
+}
+
+fn convert<E: Into<Box<dyn Error + Send + Sync>>>(
+    poll: Poll<Result<(), E>>,
+) -> Poll<io::Result<()>> {
+    match poll {
+        Poll::Pending => Poll::Pending,
+        Poll::Ready(Ok(())) => Poll::Ready(Ok(())),
+        Poll::Ready(Err(e)) => Poll::Ready(Err(io::Error::new(io::ErrorKind::Other, e))),
+    }
+}
+
+impl<S, Item, ItemDecoder, SinkItem, SinkItemEncoder> From<(S, ItemDecoder, SinkItemEncoder)>
+    for Transport<S, Item, ItemDecoder, SinkItem, SinkItemEncoder>
+where
+    S: AsyncWrite + AsyncRead,
+    Item: for<'de> Deserialize<'de>,
+    ItemDecoder: Deserializer<Item>,
+    SinkItem: Serialize,
+    SinkItemEncoder: Serializer<SinkItem>,
+{
+    fn from((inner, decoder, encoder): (S, ItemDecoder, SinkItemEncoder)) -> Self {
+        Transport {
+            inner: FramedRead::new(
+                FramedWrite::new(Framed::new(inner, LengthDelimitedCodec::new()), encoder),
+                decoder,
+            ),
+        }
+    }
+}
+
+#[cfg(feature = "tcp")]
+#[cfg_attr(docsrs, doc(cfg(feature = "tcp")))]
+/// TCP support for generic transport using Tokio.
+pub mod tcp {
+    use {
+        super::*,
+        futures::ready,
+        std::{marker::PhantomData, net::SocketAddr},
+        tokio::net::{TcpListener, TcpStream, ToSocketAddrs},
+    };
+
+    mod private {
+        use super::*;
+
+        pub trait Sealed {}
+
+        impl<Item, ItemDecoder, SinkItem, SinkItemEncoder> Sealed
+            for Transport<TcpStream, Item, ItemDecoder, SinkItem, SinkItemEncoder>
+        {
+        }
+    }
+
+    /// TCP extensions for the generic transport.
+    pub trait TransportExt: private::Sealed {
```

Comment:
Nit: you don't need this trait. Because it's in the same crate as Transport, you can provide an impl directly: `impl<Item, ItemDecoder, SinkItem, SinkItemEncoder> Transport<TcpStream, Item, ItemDecoder, SinkItem, SinkItemEncoder>{...}`